libvips/conversion/autorot.c

```diff
@@ -170,27 +146,93 @@ vips_autorot_build( VipsObject *object )
 {
 	VipsConversion *conversion = VIPS_CONVERSION( object );
 	VipsAutorot *autorot = (VipsAutorot *) object;
-	VipsImage **t = (VipsImage **) vips_object_local_array( object, 2 );
+	VipsImage **t = (VipsImage **) vips_object_local_array( object, 3 );
 
 	if( VIPS_OBJECT_CLASS( vips_autorot_parent_class )->build( object ) )
 		return( -1 );
-
-	g_object_set( object, 
-		"angle", vips_autorot_get_angle( autorot->in ),
-		NULL ); 
-
-	if( vips_autorot_handled( autorot->in ) ) {
-		if( vips_rot( autorot->in, &t[0], autorot->angle, NULL ) ||
-			vips_copy( t[0], &t[1], NULL ) )
-			return( -1 );
-		vips_autorot_remove_angle( t[1] ); 
-	}
-	else {
-		if( vips_copy( autorot->in, &t[1], NULL ) )
-			return( -1 );
-	}
-
-	if( vips_image_write( t[1], conversion->out ) )
+	
+    int orientation = 0;
+    VipsAngle angle;
+    gboolean flip = FALSE;
+    
+    if( !vips_image_get_typeof( autorot->in, VIPS_META_ORIENTATION ) ||
+        vips_image_get_int( autorot->in, VIPS_META_ORIENTATION, &orientation ) )
+        orientation = 1;
+
+    switch( orientation ) {
+
+    case 2:
+        angle = VIPS_ANGLE_D0;
+        flip = TRUE;
+        break;
+
+    case 4:
+        flip = TRUE;
+
+    case 3:
+        angle = VIPS_ANGLE_D180;
+        break;
+        
+    case 5:
+        flip = TRUE;
+
+    case 6:
+        angle = VIPS_ANGLE_D90;
+        break;
+        
+    case 7:
+        flip = TRUE;
+
+    case 8:
+        angle = VIPS_ANGLE_D270;
+        break;
+        
+    default:
+        angle = VIPS_ANGLE_D0;
+        flip = FALSE;
+        break;
+        
+    }
+
+    g_object_set( object,
+                  "angle", angle,
+                  NULL );
+
+    g_object_set( object,
+                  "flip", flip,
+                  NULL );
+
+    if( angle != VIPS_ANGLE_D0 && flip) {
```

Comment:
This if/else can be simplified and made more efficient: 
1. Define an 'out' pointer, which first points to autorot->in
1. If rotate needed, rotate from out into t[0], then set out = t[0]
2. If flip needed, flip from out to t[1], then set out = t[1]
3. don't copy, it's unnecessary here
4. Use 'out' instead of t[2] when calling vips_autorot_remove_angle() and vips_image_write()

You don't have to worry about memory deallocation, t[0] and t[1] will each be deallocated as before, and having a regular C pointer point to one of them will have no effect (if you connected it to the parent object like t is connected then you'd have a problem).
Then you don't need to resize the t array, leave it at size 2

Reply:
4. copy is needed due concurrency issues when working with the headers, this is a given

From that the rest is derived

I had an ref error when trying to use t[0] as in